plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/RocksDBKeyValueStorageFactory.java

```diff
@@ -149,55 +153,71 @@ public KeyValueStorage create(
-        {
-          segmentedStorage = null;
-          if (unsegmentedStorage == null) {
-            unsegmentedStorage =
-                new RocksDBKeyValueStorage(
-                    rocksDBConfiguration, metricsSystem, rocksDBMetricsFactory);
-          }
-          return unsegmentedStorage;
+      case 0 -> {
+        segmentedStorage = null;
+        if (unsegmentedStorage == null) {
+          unsegmentedStorage =
+              new RocksDBKeyValueStorage(
+                  rocksDBConfiguration, metricsSystem, rocksDBMetricsFactory);
         }
-      case 1:
-      case 2:
-        {
-          unsegmentedStorage = null;
-          if (segmentedStorage == null) {
-            final List<SegmentIdentifier> segmentsForVersion =
-                segments.stream()
-                    .filter(segmentId -> segmentId.includeInDatabaseVersion(databaseVersion))
-                    .collect(Collectors.toList());
-
+        return unsegmentedStorage;
+      }
+      case 1, 2 -> {
+        unsegmentedStorage = null;
+        if (segmentedStorage == null) {
+          final List<SegmentIdentifier> segmentsForVersion =
+              segments.stream()
+                  .filter(segmentId -> segmentId.includeInDatabaseVersion(databaseVersion))
+                  .collect(Collectors.toList());
+          if (isForestStorageFormat) {
+            LOG.debug("FOREST mode detected, using TransactionDB.");
             segmentedStorage =
-                new RocksDBColumnarKeyValueStorage(
+                new TransactionDBRocksDBColumnarKeyValueStorage(
+                    rocksDBConfiguration,
+                    segmentsForVersion,
+                    ignorableSegments,
+                    metricsSystem,
+                    rocksDBMetricsFactory);
+          } else {
+            LOG.debug("Using OptimisticTransactionDB.");
+            segmentedStorage =
+                new OptimisticRocksDBColumnarKeyValueStorage(
                     rocksDBConfiguration,
                     segmentsForVersion,
                     ignorableSegments,
                     metricsSystem,
                     rocksDBMetricsFactory);
           }
-          final RocksDbSegmentIdentifier rocksSegment =
-              segmentedStorage.getSegmentIdentifierByName(segment);
-          return new SegmentedKeyValueStorageAdapter<>(
-              segment, segmentedStorage, () -> segmentedStorage.takeSnapshot(rocksSegment));
         }
-      default:
-        {
-          throw new IllegalStateException(
-              String.format(
-                  "Developer error: A supported database version (%d) was detected but there is no associated creation logic.",
-                  databaseVersion));
+
+        final RocksDbSegmentIdentifier rocksSegment =
+            segmentedStorage.getSegmentIdentifierByName(segment);
+
+        if (isForestStorageFormat) {
+          return new NonSnappableSegmentedKeyValueStorageAdapter<>(segment, segmentedStorage);
```

Comment:
not to be too naming nitpicky, but IMO we should define these in the 'positive', like SnappableSegmentedKeyValueStorageAdapter and Segmented...Adapter

Reply:
done